src/games/strategy/engine/framework/map/download/MapDownloadProgressPanel.java

```diff
@@ -0,0 +1,125 @@
+    downloadCoordinator = new DownloadCoordinator();
+    downloadCoordinator.setAllDownloadCompleteAction(() -> {
+      parent.setVisible(false);
+      parent.dispose();
+    });
+  }
+
+
+  public void cancel() {
+    downloadCoordinator.cancelDownloads();
+  }
+
+
+  public void download(List<DownloadFileDescription> newDownloads) {
+    List<DownloadFileDescription> brandNewDownloads = Lists.newArrayList();
+    for (DownloadFileDescription download : newDownloads) {
+      if (!downloadList.contains(download) && !download.isDummyUrl() && !download.getMapName().isEmpty()) {
+        brandNewDownloads.add(download);
+      }
+    }
+    newDownloads = brandNewDownloads;
+
+    if (newDownloads.isEmpty()) {
+      return;
+    }
+
+    int itemCount = newDownloads.size() + downloadList.size();
+    this.removeAll();
+    add(SwingComponents.horizontalJPanel(labelGrid, progressGrid));
+
+    labelGrid.setLayout(new GridLayout(itemCount, 1));
+    progressGrid.setLayout(new GridLayout(itemCount, 1));
+
+
+
+    for (DownloadFileDescription download : newDownloads) {
+      if (download.isDummyUrl() || download.getMapName().isEmpty()) {
+        continue;
+      }
+      // space at the end of the label so the text does not end right at the progress bar
+      labels.put(download, new JLabel(download.getMapName() + "  "));
+      JProgressBar progressBar = new JProgressBar();
+      progressBar.setStringPainted(true);
+      progressBar.setToolTipText("Installing to: " + download.getInstallLocation().getAbsolutePath());
+
+      progressBars.put(download, progressBar);
+    }
+
+    for (int i = newDownloads.size() - 1; i >= 0; i--) {
+      // add new downoads to the head of the list, this will allow the user to see newly added items directly,
+      // rather than having to scroll down to see new items.
+      downloadList.add(0, newDownloads.get(i));
+    }
+
+    for (DownloadFileDescription download : downloadList) {
+      labelGrid.add(labels.get(download));
+      progressGrid.add(progressBars.get(download));
+    }
+
+    revalidate();
+    repaint();
+
+    for (DownloadFileDescription download : newDownloads) {
+      if (download.isDummyUrl() || download.getMapName().isEmpty()) {
+        continue;
+      }
+      final JProgressBar progressBar = progressBars.get(download);
+      Consumer<Integer> progressListener = s -> {
+        SwingUtilities.invokeLater(() -> progressBar.setValue(s));
+      };
+      Runnable completionListener = () -> {
+        SwingUtilities.invokeLater(() -> progressBar.setValue(progressBar.getMaximum()));
+      };
+
+      (new Thread(() -> {
+        int length = DownloadUtils.getDownloadLength(download.newURL());
+        SwingUtilities.invokeLater(() -> {
+          progressBar.setMinimum(0);
+          progressBar.setMaximum(length);
+        });
+        downloadCoordinator.accept(download, progressListener, completionListener);
+      })).start();
```

Comment:
```
+      (new Thread(() -> {
+        int length = DownloadUtils.getDownloadLength(download.newURL());
+        SwingUtilities.invokeLater(() -> {
+          progressBar.setMinimum(0);
+          progressBar.setMaximum(length);
+        });
+        downloadCoordinator.accept(download, progressListener, completionListener);
+      })).start();
```

The non-strung together version does not help very much:
```
+      Thread t = new Thread(() -> {
+        int length = DownloadUtils.getDownloadLength(download.newURL());
+        SwingUtilities.invokeLater(() -> {
+          progressBar.setMinimum(0);
+          progressBar.setMaximum(length);
+        });
+        downloadCoordinator.accept(download, progressListener, completionListener);
+      }));
        t..start();
```
We could extract some of these objects or code blocks to helper methods/variables, but it's arguably overkill for just the one value.